network/manager.go

```diff
@@ -0,0 +1,569 @@
+	mc   *nkn.MultiClient
+	c    *admin.Client
+
+	sync.RWMutex
+	networkData *networkData // persisted data that will be saved to disk
+}
+
+var manager *Manager
+
+func NewManager(opts *config.Opts) (*Manager, error) {
+	if manager != nil {
+		return manager, nil
+	}
+
+	manager = &Manager{opts: opts}
+	err := manager.loadNetworkData()
+	if err != nil {
+		log.Println("Failed to loadNetworkData:", err)
+		return nil, err
+	}
+
+	mc, err := GetMultiClient(opts)
+	if err != nil {
+		log.Println("Failed to get multi client:", err)
+		return nil, err
+	}
+	manager.mc = mc
+	manager.c = &admin.Client{MultiClient: mc}
+
+	return manager, nil
+}
+
+func (m *Manager) StartManager() error {
+	log.Println("nConnect network manager is listening at:", m.mc.Address())
+
+	for {
+		msg := <-m.mc.OnMessage.C
+		resp, err := m.handleRequest(msg)
+		if err != nil {
+			log.Println("nConnect manager handle request error", err)
+			continue
+		}
+		if m.opts.Verbose {
+			log.Printf("nConnect manager send response: %+v\n", resp)
+		}
+
+		b, err := json.Marshal(resp)
+		if err != nil {
+			log.Println("nConnect manager json.Marshal resp error", err)
+			continue
+		}
+
+		err = msg.Reply(b)
+		if err != nil {
+			log.Println("nConnect manager msg.Reply error", err)
+			continue
+		}
+	}
+}
+
+func (m *Manager) handleRequest(msg *nkn.Message) (*managerToMember, error) {
+	req := &memberToManager{}
+	err := json.Unmarshal(msg.Data, req)
+	if err != nil {
+		log.Println("nConnect manager Unmarshal request error:", err)
+		return nil, err
+	}
+
+	if m.opts.Verbose {
+		log.Printf("nConnect manager received request from %s: %+v\n", msg.Src, req)
+	}
+
+	var node *nodeInfo
+	resp := &managerToMember{}
+	resp.MsgType = req.MsgType
+
+	switch req.MsgType {
+	case JOIN_NETWORK:
+		resp.NetworkInfo = m.networkData.NetworkInfo
+		node, err = m.JoinNetwork(msg.Src, req.Name)
+		if node != nil {
+			resp.NodeInfo = append(resp.NodeInfo, node)
+		}
+
+	case LEAVE_NETWORK:
+		err = m.LeaveNetwork(msg.Src, req.Name)
+
+	case GET_MY_INFO:
+		resp.NetworkInfo = m.networkData.NetworkInfo
+		if n := m.GetNodeInfo(msg.Src); n != nil {
+			resp.NodeInfo = append(resp.NodeInfo, n)
+		} else {
+			err = errNodeNotFound
+		}
+
+	case GET_NODES_I_ACCEPT:
+		list := m.GetAcceptNodes(msg.Src)
+		resp.NodeInfo = list
+
+	case GET_NODES_I_CAN_ACCESS:
+		list := m.GetNodesICanAccess(msg.Src)
+		resp.NodeInfo = list
+
+	case UPDATE_MY_INFO:
+		resp.NetworkInfo = m.networkData.NetworkInfo
+		err = m.UpdateName(msg.Src, req.Name)
+
+	default:
+		return nil, fmt.Errorf("nConnect manager got unknown message type: %v", req.MsgType)
+	}
+
+	if err != nil {
+		resp.Err = err
+	}
+
+	return resp, nil
+}
+
+func (m *Manager) JoinNetwork(address, name string) (*nodeInfo, error) {
+	if node, ok := m.networkData.Member[address]; ok {
+		node.LastSeen = time.Now()
+		m.networkData.Member[address] = node
+		if err := m.saveNetworkData(); err != nil {
+			return nil, err
+		}
+		noti := &managerToMember{

```

Comment:
what does `noti` stands for?

Reply:
Sorry, it is abbreviation for `notification`